### PR TITLE
User API support

### PR DESCRIPTION
**Pull request by heldersepu**
Fix last part of #41 .

Tested with:
https://github.com/heldersepu/hs-scripts/blob/master/TerraForm/sumologic/main.tf

```
heldersepu@OWL:~/code/hs-scripts/TerraForm/sumologic$ export TF_LOG=; terraform init; terraform apply -auto-approve; terraform destroy -auto-approve

Initializing provider plugins...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
sumologic_role.example_role: Creating...
  description:    "" => "Testing resource sumologic_role"
  destroy:        "" => "true"
  lookup_by_name: "" => "false"
  name:           "" => "TestRole123"
sumologic_collector.example_collector: Creating...
  category:       "" => "my/source/category2"
  destroy:        "" => "true"
  lookup_by_name: "" => "false"
  name:           "" => "Test Collector123"
  timezone:       "" => "Etc/UTC"
sumologic_collector.example_collector: Creation complete after 2s (ID: 159858433)
sumologic_role.example_role: Creation complete after 2s (ID: 0000000000967720)
sumologic_user.example_user2: Creating...
  active:     "" => "true"
  email:      "" => "jane.smith@gmail.com"
  first_name: "" => "Jane"
  last_name:  "" => "Smith"
  role_ids.#: "0" => "1"
  role_ids.0: "" => "0000000000967720"
sumologic_user.example_user1: Creating...
  active:     "" => "false"
  email:      "" => "jon.doe@gmail.com"
  first_name: "" => "Jon"
  last_name:  "" => "Doe"
  role_ids.#: "0" => "1"
  role_ids.0: "" => "0000000000967720"
sumologic_user.example_user2: Creation complete after 6s (ID: 0000000000967757)
sumologic_user.example_user1: Creation complete after 7s (ID: 000000000096775F)

Apply complete! Resources: 4 added, 0 changed, 0 destroyed.
sumologic_role.example_role: Refreshing state... (ID: 0000000000967720)
sumologic_collector.example_collector: Refreshing state... (ID: 159858433)
sumologic_user.example_user2: Refreshing state... (ID: 0000000000967757)
sumologic_user.example_user1: Refreshing state... (ID: 000000000096775F)
sumologic_user.example_user1: Destroying... (ID: 000000000096775F)
sumologic_user.example_user2: Destroying... (ID: 0000000000967757)
sumologic_collector.example_collector: Destroying... (ID: 159858433)
sumologic_collector.example_collector: Destruction complete after 1s
sumologic_user.example_user1: Destruction complete after 6s
^[[A^[[D^[[Asumologic_user.example_user2: Still destroying... (ID: 0000000000967757, 10s elapsed)
sumologic_user.example_user2: Destruction complete after 11s
sumologic_role.example_role: Destroying... (ID: 0000000000967720)
sumologic_role.example_role: Destruction complete after 0s

Destroy complete! Resources: 4 destroyed.
```

### CODE REVIEW

**Comment by heldersepu** [comment]
@mtanda I see you added the support for Roles, on this PR i'm getting an error:
`Status code 404 - unexpected end of JSON input` 
The TF code to reproduce it is here:
https://github.com/heldersepu/hs-scripts/blob/master/TerraForm/sumologic/main.tf

**Comment by mtanda** [comment]
@heldersepu Hi, I try to replicate the issue.
And, I should say sorry, I accidentally apply to your account because I just copy your tf file, and apply it. I hope my apply doesn't break your account...

**Comment by mtanda** [comment]
`roles` also has `users` field.
I guess `roles` should be updated when updating `users`.
https://api.us2.sumologic.com/docs/#operation/updateRole

**Comment by mtanda** [comment]
I can register users to roles like following input.

```
resource "sumologic_role" "Foo" {
  name = "Foo"
  users = [
    "00000000000AAA01",
    "00000000000AAA02",
    "00000000000AAA03"
  ]
  capabilities = [
  ]
}
```

So if users is created correctly, it should work.

According to document, users need to create with at least one role.
In my org, we use SAML, and assign default role if user's info doesn't match predefined role.
We don't use Sumo Logic UI or User API to create user...

**Comment by mtanda** [comment]
More simple user creation input, it seems to failed in terraform, but user is created.
```
❯ cat *.tf
provider "sumologic" {
  environment = "jp"
}

resource "sumologic_user" "example_user" {
  first_name = "Jon"
  last_name  = "Doe"
  email      = "jd@gmail.com"
  role_ids   = [
    "000000000008AA2E"
  ]
}

❯ /usr/local/bin/terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

sumologic_user.example_user: Refreshing state... (ID: 0000000000164389)

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  + sumologic_user.example_user
      id:         <computed>
      destroy:    "true"
      email:      "jd@gmail.com"
      first_name: "Jon"
      last_name:  "Doe"
      role_ids.#: "1"
      role_ids.0: "000000000008AA2E"


Plan: 1 to add, 0 to change, 0 to destroy.

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.


❯ /usr/local/bin/terraform apply
sumologic_user.example_user: Refreshing state... (ID: 0000000000164389)

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  + sumologic_user.example_user
      id:         <computed>
      destroy:    "true"
      email:      "jd@gmail.com"
      first_name: "Jon"
      last_name:  "Doe"
      role_ids.#: "1"
      role_ids.0: "000000000008AA2E"


Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

sumologic_user.example_user: Creating...
  destroy:    "" => "true"
  email:      "" => "jd@gmail.com"
  first_name: "" => "Jon"
  last_name:  "" => "Doe"
  role_ids.#: "0" => "1"
  role_ids.0: "" => "000000000008AA2E"

Error: Error applying plan:

1 error occurred:
        * sumologic_user.example_user: 1 error occurred:
        * sumologic_user.example_user: Status code 404 - unexpected end of JSON input





Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.



❯ curl -s -u "${SUMOLOGIC_ACCESSID}:${SUMOLOGIC_ACCESSKEY}" -X GET 'https://api.jp.sumologic.com/api/v1/users'| jq .|grep 'jd@'
      "email": "jd@gmail.com",
```

**Comment by mtanda** [comment]
Would you try to check the simple input is success in your account?

**Comment by heldersepu** [comment]
Hello @mtanda  you mentioned

> roles also has users field.
> I guess roles should be updated when updating users.
> https://api.us2.sumologic.com/docs/#operation/updateRole

Yes but the users in the roles is not required, and in the User it is: 
https://api.sumologic.com/docs/#operation/updateUser

My idea here was create the roles first and then the users 
I guess I have to add a `lifecycle` config ignoring changes
https://www.terraform.io/docs/configuration/resources.html#ignore_changes

...and no worries about this account this is just a trial account

**Comment by heldersepu** [comment]
I just checked the "simple input" but that expects that the role is already there.
That will not work on a real life situation, everything must be created by terraform...

I made changes to my code adding:
 `lifecycle { ignore_changes = ["users"] }` 
on the `resource "sumologic_role"`
That way the second and following times we run the apply we wont get the updates...

## But the initial error remains ... 
We still get the following on a clean run:
`sumologic_user.example_user: Status code 404 - unexpected end of JSON input`
That is the error I'm trying to get to the bottom of, as far as I can tell the user is created successfully with the correct role

**Comment by heldersepu** [comment]
Figure out what I was doing wrong... 
The Put does not have the same schema as the Post
https://api.sumologic.com/docs/#operation/updateUser

File: sumologic/resource_sumologic_user.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+	return resourceSumologicUserRead(d, meta)
+}
+
+func resourceSumologicUserExists(d *schema.ResourceData, meta interface{}) (bool, error) {
+	c := meta.(*Client)
+
+	_, err := c.GetUser(d.Id())
+
+	return err == nil, nil
```
Comment (horjulf) [inline sumologic/resource_sumologic_user.go]: In this case, API errors will represent a missing user.
Validate the user pointer and return `err` directly.

File: sumologic/resource_sumologic_user.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+		id, err := c.CreateUser(user)
+
+		if err != nil {
+			return err
+		}
+
+		d.SetId(id)
+	}
+
+	return resourceSumologicUserUpdate(d, meta)
```
[the remark this answers is not in the document]
Comment (heldersepu) [reply sumologic/resource_sumologic_user.go]: Yes update sets the isActive flag:
https://api.sumologic.com/docs/#operation/updateUser

**Comment by heldersepu** [comment]
@frankreno Any chance we can merge this PR?

**Comment by frankreno** [comment]
@heldersepu  - apologies, I will review this next week.

**Comment by heldersepu** [comment]
@frankreno maybe this week?

**Comment by heldersepu** [comment]
@frankreno sounds good, do you guys use any tools to generate base docs?

**Comment by frankreno** [comment]
Unfortunately, we do not right now. It also looks like we are missing roles I just realized.  @mtanda can you submit a PR for the role changes you made at some point? The recently merged Ingest Budgets additions have a good exampe - https://github.com/SumoLogic/sumologic-terraform-provider/pull/57/files

**Comment by mtanda** [comment]
@frankreno I'll send a PR for roles :-)

**Comment by heldersepu** [comment]
@frankreno doc added:
https://github.com/SumoLogic/sumologic-terraform-provider/blob/ae3bc6be29d39e68190a61021c56fdc145a27c42/docs/r/sumologic_user.md

**Comment by frankreno** [comment]
Thanks @heldersepu ! LGTM

**Pull request merged**